cpp/src/arrow/dataset/file_base.h

```diff
@@ -350,23 +351,25 @@ class ARROW_DS_EXPORT FileWriter {
 /// \brief Options for writing a dataset.
 struct ARROW_DS_EXPORT FileSystemDatasetWriteOptions {
   /// Options for individual fragment writing.
-  std::shared_ptr<FileWriteOptions> file_write_options;
+  std::shared_ptr<FileWriteOptions> file_write_options =
+      CsvFileFormat().DefaultWriteOptions();
 
   /// FileSystem into which a dataset will be written.
-  std::shared_ptr<fs::FileSystem> filesystem;
+  std::shared_ptr<fs::FileSystem> filesystem =
+      std::make_shared<arrow::fs::LocalFileSystem>();
 
   /// Root directory into which the dataset will be written.
   std::string base_dir;
 
   /// Partitioning used to generate fragment paths.
-  std::shared_ptr<Partitioning> partitioning;
+  std::shared_ptr<Partitioning> partitioning = Partitioning::Default();
 
   /// Maximum number of partitions any batch may be written into, default is 1K.
   int max_partitions = 1024;
 
   /// Template string used to generate fragment basenames.
   /// {i} will be replaced by an auto incremented integer.
-  std::string basename_template;
+  std::string basename_template = "data_{i}.arrow";
```

Comment:
The extension should be based on the format.  We should add a `const std::string& default_extension()` method to `FileFormat`.  Then we should default this to the empty string.  Then, in the dataset writer, if this is an empty string, we should use `"part-{i}." + format.default_extension()`.  This mimics what is done in python (and we could probably remove the python logic too).  Then we should update the docstring for this field to reflect this behavior.

Reply:
`FileFormat` has a funtion called `type_name()` which currentyle is returning the `dataset-file-formats`, so I think 
 `default_extension()` is not necessary

Reply:
`type_name()` almost works.  However, for `IpcFileFormat` the `type_name` is `ipc` and the extension should be `arrow`.  We could probably update the `type_name` to be `arrow` though.  It would technically be a backwards incompatible change but I'm not sure if anyone uses this field today.